doc/adr007-flattening.md

```diff
@@ -0,0 +1,255 @@
+
+### Flattener
+
+The flattener adds missing definitions to a module by recursively fetching the lookup table and copying the referenced definition, then looking into that definition for any other names and doing the same.
+
+When adding a missing definition, if that definitions carries a namespace, we actually need to do a "deep" procedure to add that name to not only the definition name, but also to every other name appearing in it's body. This includes adding the namespace to lambda parameters and nested definitions (even though they cannot be referred), so flattening doesn't create name conflicts with them.
+
+This deep namespace-adding procedure is already done in the current implementation of flattening (see [`addNamespaceToDef`](https://github.com/informalsystems/quint/blob/main/quint/src/flattening.ts#L276C4-L276C4)), with the only difference being that we don't want to generate new ids in this case (we only create new ids in the Instance Flattener).
+
+Although the Instance Flattener requires imports and instances to be flattened, we can't flatten definitions that come from instances before running the instance flattener, since that would copy definitions that would potentially be replaced by new definitions in new modules created by the instance flattener. Therefore, we need to run the flattener twice: first not copying definitions that come from instances (checking the `importedFrom` field once again), and then, after the instance flattener, flattening anything that remains to be flattened.
+
+### Overall flattening process
+
+The flattening process is run completely for each module before proceeding to flatten the next module. The modules are assumed to be topologically sorted. The need for running this by module is that we do remove the import statements after we resolve all dependencies, but we can only ensure that all dependencies were flattened if the depending modules were completely flattened before.
+
+The whole flattening process is composed by the following steps:
+1. Run the Flattener, ignoring definitions that come from instances
+2. Remove `import` and `export` statements
+3. Run the Instance Flattener
+4. Resolve names for the module and its dependencies, obtaining a new lookup table
+5. With the new table and modules, run the Flattener again, this time for all definitions
+6. Resolve names again for all modules (using the flattened version for those that are already flattened, and the original version for those that are not). The resulting table is used as the starting point of flattening for the next module, or returned as the final table if this is the last one.
+	 - We might be able to manipulate the table inside the flattener so this is not necessary, but it is not obvious how to do that. The main problem seems to be that, since the Instance Flattener replaced instance statements with import statements, the `importFrom` and `namespaces` fields of certain definitions have to be updated.
```

Comment:
```suggestion
	 - We might be able to manipulate the table inside the flattener so this is not necessary, but it is not obvious how to do that. The main problem is that, since the Instance Flattener replaced instance statements with import statements, the `importFrom` and `namespaces` fields of certain definitions have to be updated.
```

Reply:
I haven't validated this hypothesis, so I'd rather leave it as "seems to be"

Reply:
fair enough